### PR TITLE
Strengthen the checking for deploymentAvailable and rolledPodsAreRead…

### PR DESCRIPTION
**Pull request by Hungrylion2019**
1. rolledPodsAreReady
-  add the UpdatedReplicas and UnavailableReplicas judge to be sure that we update deployement ok

2. deployment
- we set maxUnavailable: 2；if some other pod is not good，it will still think the deployment is ok
- I think it's not enough to be sure the deployment is healthy when some pods occured to error

Signed-off-by: zhicheng.li <zhicheng.li@trip.com>

### CODE REVIEW

**Comment by Hungrylion2019** [comment]
@integrii

File: cmd/deployment-check/deployment.go [hunk 1]
```diff
@@ -692,7 +693,10 @@ func deploymentAvailable(deployment *v1.Deployment) bool {
 	for _, condition := range deployment.Status.Conditions {
 		if condition.Type == v1.DeploymentAvailable && condition.Status == corev1.ConditionTrue {
 			log.Infoln("Deployment is reporting", condition.Type, "with", condition.Status+".")
-			return true
+			if deployment.Status.Replicas == int32(checkDeploymentReplicas) && deployment.Status.AvailableReplicas == int32(checkDeploymentReplicas) && deployment.Status.ReadyReplicas == int32(checkDeploymentReplicas) && deployment.Status.ObservedGeneration == 1 {
+				log.Infoln("The ", deployment.Status.AvailableReplicas, " pods of the deployment  are ready and available")
```
Comment (integrii) [inline cmd/deployment-check/deployment.go:697]:
```suggestion
				log.Infoln(deployment.Status.AvailableReplicas, "deployment pods are ready and available.")
```

File: cmd/deployment-check/deployment.go [hunk 2]
```diff
@@ -683,7 +683,8 @@ func rollingUpdateComplete(ctx context.Context, statuses map[string]bool, oldPod
 // rolledPodsAreReady checks if a deployments pods have been updated and are available.
 // Returns true if all replicas are up, ready, and the deployment generation is greater than 1.
 func rolledPodsAreReady(d *v1.Deployment) bool {
-	return d.Status.Replicas == int32(checkDeploymentReplicas) && d.Status.AvailableReplicas == int32(checkDeploymentReplicas) && d.Status.ReadyReplicas == int32(checkDeploymentReplicas) && d.Status.ObservedGeneration > 1
+	return d.Status.Replicas == int32(checkDeploymentReplicas) && d.Status.UpdatedReplicas == int32(checkDeploymentReplicas) && d.Status.AvailableReplicas == int32(checkDeploymentReplicas) && d.Status.ReadyReplicas == int32(checkDeploymentReplicas) && d.Status.UnavailableReplicas < 1 &&
```
Comment (integrii) [inline cmd/deployment-check/deployment.go:686]:
```suggestion
	return d.Status.Replicas == int32(checkDeploymentReplicas) &&
	d.Status.UpdatedReplicas == int32(checkDeploymentReplicas) &&
	d.Status.AvailableReplicas == int32(checkDeploymentReplicas) &&
	d.Status.ReadyReplicas == int32(checkDeploymentReplicas) &&
	d.Status.UnavailableReplicas < 1 &&
	d.Status.ObservedGeneration > 1
```

File: cmd/deployment-check/deployment.go [hunk 3]
```diff
@@ -683,7 +683,8 @@ func rollingUpdateComplete(ctx context.Context, statuses map[string]bool, oldPod
...
+		d.Status.ObservedGeneration > 1
```
Comment (integrii) [inline cmd/deployment-check/deployment.go:687]:
```suggestion
```

File: cmd/deployment-check/deployment.go [hunk 4]
```diff
@@ -692,7 +693,10 @@ func deploymentAvailable(deployment *v1.Deployment) bool {
 	for _, condition := range deployment.Status.Conditions {
 		if condition.Type == v1.DeploymentAvailable && condition.Status == corev1.ConditionTrue {
 			log.Infoln("Deployment is reporting", condition.Type, "with", condition.Status+".")
-			return true
+			if deployment.Status.Replicas == int32(checkDeploymentReplicas) && deployment.Status.AvailableReplicas == int32(checkDeploymentReplicas) && deployment.Status.ReadyReplicas == int32(checkDeploymentReplicas) && deployment.Status.ObservedGeneration == 1 {
```
Comment (integrii) [inline cmd/deployment-check/deployment.go:696]:
```suggestion
			if deployment.Status.Replicas == int32(checkDeploymentReplicas) &&
			deployment.Status.AvailableReplicas == int32(checkDeploymentReplicas) &&
			deployment.Status.ReadyReplicas == int32(checkDeploymentReplicas) &&
			deployment.Status.ObservedGeneration == 1 {
```

**Review by integrii — CHANGES_REQUESTED** [review]
Hey @Hungrylion2019 !  It's been a long time since I have seen your code.  This should be fine to do.  I just have a couple small syntax suggestions for you to look at...

**Comment by Hungrylion2019** [comment]
> Hey @Hungrylion2019 ! It's been a long time since I have seen your code. This should be fine to do. I just have a couple small syntax suggestions for you to look at...
- Firstly,tks for your review.
- Actually, I have many changes for the kuberhealthy code. But the project has landed and completed, then I'm working for a new projetct aboutt node automatically upgraded、online、offline、switching. 
- I will try my best to update the code from local to remote, and some codes maybe be committed In the name of the company [ctripcloud](https://github.com/ctripcloud)

**Comment by integrii** [comment]
@Hungrylion2019 there were build fixes in master.  Can you merge the latest master and push it to your branch please?  That should make builds pass on this branch...

**Comment by Hungrylion2019** [comment]
@integrii  pls merge this PR，it  just pass the test after I add the signature about your review suggestions.
By the way, I still can't push the code into  the repository without fork mode.

**Pull request merged**